include/IECore/Ramp.h

```diff
@@ -0,0 +1,153 @@
+	Constant = 4,
+};
+
+/// A Ramp represents a spline-like curve as it is represented in a simple UI: with a set of independent
+/// control points, and an interpolation type selected from RampInterpolation.
+///
+/// Rather than storing the lower level IECore::Spline*, we now store this Ramp type in shader networks,
+/// and only convert to the lower level class with the evaluator() function when evaluation is needed.
+template<typename X, typename Y>
+class IECORE_EXPORT Ramp
+{
+
+public :
+
+	typedef X XType;
+	typedef Y YType;
+
+	typedef CubicBasis<XType> Basis;
+	typedef std::multimap<X, Y> PointContainer;
+	typedef typename PointContainer::value_type Point;
+
+
+
+	Ramp() : interpolation( RampInterpolation::CatmullRom )
+	{
+	}
+
+	Ramp( const PointContainer &p, RampInterpolation i )
+		: points( p ), interpolation( i )
+	{
+	}
+
+	PointContainer points;
+	RampInterpolation interpolation;
+
+
+	// Convert to Cortex Spline
+	// In the future, IECore::Spline may be replaced with IECore::SplineEvaluator, and this
+	// function would be the only way to setup one.
+	IECore::Spline<X, Y> evaluator() const;
+
+	// In order to write shader parameters to renderers or USD, we need to convert Ramps to a convention
+	// that can be stored in raw OSL. Based on the arguments to the OSL spline() function, we've come up
+	// with a convention for storing a string basis and separate position and value arrays, where the
+	// the position and value arrays contain any duplicated end points necessary for the resulting OSL
+	// spline to cover the full range of the ramp.
+	//
+	// These functions convert to and from this convention ( and can also be used for converting from other
+	// similar conventions with some pre-processing ).
+	//
+	// NOTE: Some aspects of the convention we've picked are not actually very universal. We haven't
+	// found any shader's in the wild that actually expect duplicated end point ( aside from Gaffer's
+	// shader library ). Other shader authors seem to be implementing their own spline functions that don't
+	// require duplicated endpoints ( 3delight ), or put code to perform the duplication in their shader
+	// before calling spline ( PRMan ). We've stuck to duplicating the end points because unnecessarily
+	// duplicating endpoints doesn't cause any problems other than being a minor waste of performance.
+	// It seems like there's an argument that the best approach would be to put the endpoint duplication
+	// inside the shader like PRMan, but this is working fine for now.
+	void fromOSL( const std::string &basis, const std::vector<X> &positions, const std::vector<Y> &values, const std::string &identifier );
+	void toOSL( std::string &basis, std::vector<X> &positions, std::vector<Y> &values ) const;
+
+	// If there are connections to a Ramp, and we convert it to the OSL convention with duplicated endpoints,
+	// then the connections will need to be offset to match - this defines the offset.
+	int oslAdaptorOffset() const;
```

Comment:
I recall we spent a little while debating if the OSL conversions should be provided by members of this class or by ShaderNetworkAlgo, and we went for this one because it had the monotone->bezier code in it already. But this `oslAdaptorOffset()` thing is new to me, and makes me think that the API might be cleaner if the OSL conversions weren't in here. How doable would that be?